testing.md

```diff
@@ -34,6 +34,82 @@ A piece of code is not done until it's tested.
+
+Things you should keep in mind while writing tests for a codebase.
+
+### Write Complete Testing Suites
+
+Testing suites should attempt to be exhaustive in verifying the code does what is expected. None of
+these tests should require any manual work, other than executing the initial command to run the
+tests.
+
+An exhaustive test suite implies that you should be confident of merging and deploying your changes
+once all your tests pass in your continuous integration environment. Any manual steps that are needed
+to verify code changes should be encoded as automated tests that run in your continuous integration
+environment.  As an example, if to test your code, you bring up an instance of your service and
+manually send it an example request, then this flow should be encoded as an integration test.
+
+### Test Deterministically
+
+Tests should be deterministic. That means that if you run a test multiple times without changing the
+code, it should produce the same results. A test that sometimes fails and sometimes passes is not
+consistent, and therefore can't be relied on to give you confidence about the correctness of a
+program.
+
+Concretely, this means you should avoid relying on factors like randomness, time, or external
+services in the structure of your tests. For instance, if you are testing a button that changes a
+piece of UI, you shouldn't wait for an arbitrary number of seconds to check that the UI refreshes,
+since the time that it takes to refresh is not deterministic. Instead, you can emit some sort of
+event when the UI is finished refreshing, and make the test listen for this event.
+
+Testing non-deterministic code can also be very difficult. If you can, avoid non-deterministic
+behavior in your code. If you must have it, try to isolate it so it can be deterministically mocked
+in tests.
+
+One type of non-determinism that *is* useful for testing is generating random inputs.
+[Generative testing](http://blog.8thlight.com/connor-mendenhall/2013/10/31/check-your-work.html), in
+which a program generates different test inputs each time the tests are run, can help describe
+properties of your program that should hold for all inputs. They can make a stronger assertion about
+correctness, since, if they pass, it means your program works for many random inputs, not just a few
+representative ones. Generative tests can also come up with edge cases that you may not have thought
+of.
+
+### Prefer Readability and Ease of Code
+
+Tests should be written in a manner that prioritizes readability and ease of writing new tests and
+modifying existing ones. In particular, these should be prioritized over other typical concerns like
+abstraction and removing code duplication. If abstractions are needed, they should be created in a
+manner that preserves the logic in the test functions code without requiring many hops. Another way
+of saying this is that it should be apparent what the test is doing by just looking at the test
+code.
```

Comment:
one area where you might have a different standard in non-test code: if performance is essential, you might choose a harder to read/use library if it had the best performance

but generally agreed that we should favor developer ease-of-use in all code!